### PR TITLE
Added module for clear cache after release

### PR DESCRIPTION
**Pull request by 99percent6**
### Related issues

closes #

### Short description and why it's useful

After the releases, we often encounter the fact that some users have spontaneous bugs or something starts to work incorrectly or the new functionality has not reached them. All these problems are resolved after cleaning localstorage, indexedDB and service worker. This module is designed specifically for such cases.

### Which environment this relates to
Check your case. In case of any doubts please read about [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html)

- [x] Test version (https://test.storefrontcloud.io) - this is a new feature or improvement for Vue Storefront. I've created branch from `develop` branch and want to merge it back to `develop`
- [ ] RC version (https://next.storefrontcloud.io) - this is a stabilisation fix for Release Candidate of Vue Storefront. I've created branch from `release` branch and want to merge it back to `release`
- [ ] Stable version (https://demo.storefrontcloud.io) - this is an important fix for current stable version. I've created branch from `hotfix` or `master` branch and want to merge it back to `hotfix`

### Upgrade Notes and Changelog

- [x] No upgrade steps required (100% backward compatibility and no breaking changes)
- [ ] I've updated the [Upgrade notes](https://github.com/DivanteLtd/vue-storefront/blob/develop/docs/guide/upgrade-notes/README.md) and [Changelog](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md) on how to port existing VS sites with this new feature

**IMPORTANT NOTICE** - Remember to update `CHANGELOG.md` with description of your change

### Contribution and currently important rules acceptance

- [x] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: config/default.json [hunk 1]
```diff
@@ -525,5 +525,31 @@
...
+  },
+  "clearCache": {
+    "enabled": true,
+    "websocket": {
+      "enabled": true,
+      "endpoint": "/clear-cache",
+      "key": "clear-cache",
+      "localStorage": {
+        "enabled": true,
+        "keysForClear": ["elasticCache", "url", "attributes"]
```
Comment (pkarw) [inline config/default.json]: can we just have “keys”?

File: config/default.json [hunk 2]
```diff
@@ -525,5 +525,31 @@
...
+      },
+      "indexedDB": {
+        "enabled": true,
+        "database": "shop",
+        "storagesForClear": ["attributes", "categories", "elasticCache"]
+      },
```
Comment (pkarw) [inline config/default.json]: Same like above - I suggest just “keys” as a property name

File: core/scripts/server.ts [hunk 3]
```diff
@@ -135,6 +136,43 @@ app.use('/service-worker.js', serve('dist/service-worker.js', false, {
 app.post('/invalidate', invalidateCache)
 app.get('/invalidate', invalidateCache)
 
+if (config.clearCache && config.clearCache.enabled) {
```
Comment (pkarw) [inline core/scripts/server.ts]: I suggest moving this code into separate server module (see src/modules/compress for a reference)

File: config/default.json (hunk 2 above)
Comment (99percent6) [reply config/default.json]: All done

File: config/default.json (hunk 1 above)
Comment (99percent6) [reply config/default.json]: Yes, done

File: core/scripts/server.ts (hunk 3 above)
Comment (99percent6) [reply core/scripts/server.ts]: Did this

**Comment by filrak** [comment]
Really useful one. Just please wait with merge before https://github.com/DivanteLtd/vue-storefront/pull/3651as it contains potential breaking changes. Also how about releasing it as a separate npm package @pkarw. We aim to remove all modules from src so putting something there seems redundant

**Comment by filrak** [comment]
Really useful one. Just please wait with merge before https://github.com/DivanteLtd/vue-storefront/pull/3651as it contains potential breaking changes. Also how about releasing it as a separate npm package @pkarw. We aim to remove all modules from src so putting something there seems like adding ourselves extra work

**Comment by andrzejewsky** [comment]
@99percent6 what's the status of this one? Do you want to finish it?

**Comment by NicolasHofmairT23** [comment]
Hi there!

We have already implemented this PR in our project. @gibkigonzo helped us with this task.
it is working quite well.

But right now we noticed, that the "please reload" notification is **always** shown when you load the page initially in **firefox** (without any cached data. U can use inkognito tab to reproduce this). In chrome this does not happen.

You have any idea why or already know this issue?

**Comment by michhy** [comment]
> Hi there!
> 
> We have already implemented this PR in our project. @gibkigonzo helped us with this task.
> it is working quite well.
> 
> But right now we noticed, that the "please reload" notification is **always** shown when you load the page initially in **firefox** (without any cached data. U can use inkognito tab to reproduce this). In chrome this does not happen.
> 
> You have any idea why or already know this issue?

May I know how to enable this module? I tried to put the whole folder in src/module, and add the module in src/module/index.ts, and added the definition in local.json already..

**Comment by Fifciu** [comment]
Will any of you guys finish this PR @99percent6 , @NicolasHofmairT23 , @gibkigonzo ?

**Comment by bloodf** [comment]
This is closed due to inactivity.

If you want your PR to be re-opened, please update your files, check for any conflicts.

If your PR is for Vue Storefront 1 target the new repository at https://github.com/vuestorefront/vue-storefront-1

**Pull request closed without merging**